include/RAJA/policy/cuda/atomic.hpp

```diff
@@ -25,614 +25,569 @@
+  float
 #if __CUDA_ARCH__ >= 600
-template <>
-RAJA_INLINE __device__ double cuda_atomicInc<double>(double volatile *acc)
-{
-  return ::atomicAdd((double *)acc, (double)1);
-}
+  ,
+  double
 #endif
+>;
 
-
-#if __CUDA_ARCH__ >= 200
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicDec(T volatile *acc, T val)
+template <typename T,
+          RAJA::util::enable_if_is_none_of<T, cuda_atomicSub_builtin_types>* = nullptr>
+RAJA_INLINE __device__ T cuda_atomicSub(T *acc, T value)
 {
-  // See:
-  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicdec
-  return cuda_atomic_CAS_oper(acc, [=] __device__(T old) {
-    return (((old == 0) | (old > val)) ? val : (old - 1));
+  return cuda_atomicCAS(acc, [value] (T old) {
+    return old - value;
   });
 }
 
-// 32-bit unsigned atomicDec support by CUDA
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicDec<unsigned>(unsigned volatile *acc,
-                                                    unsigned value)
+template <typename T,
+          RAJA::util::enable_if_is_any_of<T, cuda_atomicSub_via_Sub_builtin_types>* = nullptr>
+RAJA_INLINE __device__ T cuda_atomicSub(T *acc, T value)
 {
-  return ::atomicDec((unsigned *)acc, value);
+  return ::atomicSub(acc, value);
 }
 
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicDec(T volatile *acc)
+template <typename T,
+          RAJA::util::enable_if_is_any_of<T, cuda_atomicSub_via_Add_builtin_types>* = nullptr>
+RAJA_INLINE __device__ T cuda_atomicSub(T *acc, T value)
 {
-  return cuda_atomic_CAS_oper(acc,
-                                      [=] __device__(T a) { return a - 1; });
+  return ::atomicAdd(acc, -value);
 }
 
-// 32-bit signed atomicSub support by CUDA, used as backend for atomicDec
-template <>
-RAJA_INLINE __device__ int cuda_atomicDec<int>(int volatile *acc)
-{
-  return ::atomicSub((int *)acc, (int)1);
-}
 
-// 32-bit unsigned atomicSub support by CUDA, used as backend for atomicDec
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicDec<unsigned>(unsigned volatile *acc)
-{
-  return ::atomicSub((unsigned *)acc, (unsigned)1);
-}
+/*!
+ * Atomic min/max
+ */
+using cuda_atomicMinMax_builtin_types = ::camp::list<
+  int,
+  unsigned int
+#if __CUDA_ARCH__ >= 500
+  ,
+  long long int,
+  unsigned long long int
 #endif
+>;
 
 
-#if __CUDA_ARCH__ >= 200
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicAnd(T volatile *acc, T value)
+/*!
+ * Atomic min
+ */
+template <typename T,
+          RAJA::util::enable_if_is_none_of<T, cuda_atomicMinMax_builtin_types>* = nullptr>
+RAJA_INLINE __device__ T cuda_atomicMin(T *acc, T value)
 {
-  return cuda_atomic_CAS_oper(acc, [=] __device__(T a) {
-    return a & value;
-  });
+  return cuda_atomicCAS(acc,
+                        [value] (T old) {
+                          return value < old ? value : old;
+                        },
+                        [value] (T current) {
+                          return current < value;
```

Comment:
I think we can also short circuit if they are equal.

Reply:
Good call. Done.